### ISSUE TITLE
SqlServerLogin doesn't appear to be happy with some policy enforcement settings

### ISSUE THREAD

**Issue by imarambiocatan** [PowerShell/SqlServerDsc#1140]
Not my lucky week it seems with regards to playing with software
My mof file has a section like this

```
ResourceID = "[SqlServerLogin]dsc_test_Password_policy";
 **LoginPasswordPolicyEnforced = True;**
 ServerName = "[redacted]";
 InstanceName = "MSSQLSERVER";
 SourceInfo = "[redacted]";
 Name = "dsc_test";
 ModuleName = "SqlServerDsc";
 **LoginPasswordExpirationEnabled = True;**
 ModuleVersion = "11.2.0.0";

 ConfigurationName = "SQL_config";

};
```

When logged into the server, this is the result of asking the database directly about this

```powershell
PS [redacted]> Get-SqlInstance -ServerInstance localhost | get-sqllogin -name dsc_test | select Pass
wordExpirationEnabled, PasswordPolicyEnforced

PasswordExpirationEnabled PasswordPolicyEnforced
------------------------- ----------------------
                     **True                   True**
```

however, this is what dsc tells ms
```powershell
PS [redacted]> $t = Test-DscConfiguration -Path .
PS [redacted]> $t.ResourcesNotInDesiredState | where { $_.ResourceId -match "dsc_test"}

ConfigurationName    : SQL_config
DependsOn            :
ModuleName           : SqlServerDsc
ModuleVersion        : 11.2.0.0
PsDscRunAsCredential :
ResourceId           : [SqlServerLogin]dsc_test_Password_policy
SourceInfo           :  [redacted]
DurationInSeconds    : 0.078
Error                :
FinalState           :
**InDesiredState       : False**
InitialState         :
InstanceName         : dsc_test_Password_policy
RebootRequested      : False
ResourceName         : SqlServerLogin
StartDate            : 6/6/2018 4:16:40 PM
PSComputerName       : [redacted]
```

**Comment by johlju** [comment]
I updated your text with code blocks, hope you didn't mind. I thank you for finding all these issues! The account is that an SqlLogin? It only return the properties for an SqlLogin.

https://github.com/PowerShell/SqlServerDsc/blob/ca261181aaac9cfd39300ee5157f9a48b1422215/DSCResources/MSFT_SqlServerLogin/MSFT_SqlServerLogin.psm1#L55-L69

Although, I see another bug here, it should always return those properties as $false regardless of login type since we should always return all properties in the hash table. I submit an issue for that separately.

**Comment by imarambiocatan** [comment]
I don't agree. If the accounts are domain accounts, the result should be $true because domain accounts always respect password restrictions.
I was trying to use dsc to check the compliance of a pre-existing environment with a configuration standard. specifically, I was trying to use the sqlserverdsc module as much as possible for this. that's how I stumbled with all of these.
One enhancement if i might add. You should add a keyword for the instance property to force the dsc module to check the compliance in all existing instances of a server. It will make things much more flexible.
The dsc_test account was indeed a sqllogin

PS [redacted]> Get-SqlInstance -ServerInstance localhost | get-sqllogin -name "dsc_test"

Name                                          Login Type    Created               
----                                          ----------    -------               
dsc_test                                      SqlLogin      6/6/2018 3:52 PM

**Comment by johlju** [comment]
I'm not seeing `LoginType = 'SqlLogin'` in your configuration. If you add that, will it work?

Regarding the result of property `LoginPasswordPolicyEnforced` (and other properties that are unique for SQL logins) for domain accounts. I understand your point, but the resource returns a property in SQL Server, if that property exist for logins that are domain accounts and that is set to `$true` then we should return `$true`, otherwise I think we must return `$false` - best would have been to return `$null` if the property cannot be evaluated, but a Boolean cannot be set to `$null`. But if the LoginType is anything other than 'SqlLogin' then those properties should not be evaluated in the result from Get-DscConfiguration (unless mentioned above, that they actually exist as properties on domain account logins in SQL Server).

**Comment by johlju** [comment]
> One enhancement if i might add. You should add a keyword for the instance property to force the dsc module to check the compliance in all existing instances of a server. It will make things much more flexible

The property `InstanceName` is a Key, so that you can add the same login 'MyUserAccount' two to different instances. I think tou are suggesting to be able to provide a wildcard '*', so that you can add a user to all instances on the server?  It would theoretical possible. Please submit a new issue for that so the community can discuss that. 🙂

**Comment by imarambiocatan** [comment]
Yes, I know. In fact, I had some powershell code doing just that
`          $instances = Invoke-Command -ComputerName $h { 
            import-module sqlserver
            $(Get-SqlInstance -ServerInstance localhost).name
            } 
            foreach($instance in $instances)
            {
                $t = Invoke-Command -ComputerName $h { 
                    import-module sqlserver
                    Get-SqlLogin -ServerInstance $Using:instance
                    } | Where-Object { $_.logintype -eq "SqlLogin" -and $_.isdisabled -eq $false  }
                foreach($u in $t)
                {
                    $pname = $u.name + "_" + $instance + "_Password_policy"
                    $pname = $pname -replace " ","_"
                    $pname = $pname -replace "\\", "_" 
            	    SqlServerLogin $pname
	                {
                            ServerName = $h
    		                Name = $u.name
                            LoginPasswordPolicyEnforced = $True
	                        LoginPasswordExpirationEnabled = $True
                            InstanceName = $instance -replace $h, "MSSQLSERVER"
	                }#end of each instance user
                }#end of instance users
          }`
which is just crude and wrong...

**Comment by johlju** [comment]
When you got all the SQL logins, maybe you can call `Get` using [`Invoke-DscResource`](https://docs.microsoft.com/en-us/powershell/module/psdesiredstateconfiguration/invoke-dscresource?view=powershell-5.1) with just the mandatory parameters, and get the returned hash table which you might be able to use to continue evaluating.

**Comment by johlju** [comment]
You should also look at the concept ReverseDsc http://nikcharlebois.com/introducing-reverse-dsc/

**Comment by imarambiocatan** [comment]
I actually did see reverse dsc. I needed something much more dynamic. I was also having problems connecting to the target database using some of the sqlserverdsc resources.
Coupled with the things that were missing from the module, I ended up removing the use of sqlserverdsc and resorting to the script dsc resource with a lot of sqlserver cmdlets
Due to the nature of what i was doing, i needed only the get and test scripts, the set script is mostly empty
I checked for:
- the path to all the databases of all the instances (they had to be somewhere other than the system drive)
- the path to the sql server install dir
- the password policy of the sqllogin users
- the store procedures assigned to the public role
- the permissions of the guest user
- a lof ot settings from both the instance and the database

I hate SQL, but the sqlserver powershell module made things a lot easier for me. no queries at all
